pydoctor/test/test_astbuilder.py

```diff
@@ -1681,3 +1681,113 @@ def func():
     ''', systemcls=systemcls)
     outer = mod.contents['outer']
     assert not outer.contents
+
+@systemcls_param
+def test_constant_module(systemcls: Type[model.System]) -> None:
+    """
+    It can recognize constants define as module members.
+    """
+    mod = fromText('''
+    LANG = 'FR'
+    ''', systemcls=systemcls)
+    assert getattr(mod.contents['LANG'], 'kind') == model.DocumentableKind.CONSTANT
+    assert ast.literal_eval(getattr(mod.resolveName('LANG'), 'value') or '') == 'FR'
+
+@systemcls_param
+def test_constant_module_with_final(systemcls: Type[model.System]) -> None:
+    """
+    It can recognize constants defined with typing.Final
+    """
+    mod = fromText('''
+    from typing import Final
+    lang: Final = 'fr'
+    ''', systemcls=systemcls)
+    assert getattr(mod.contents['lang'], 'kind') == model.DocumentableKind.CONSTANT
+    assert ast.literal_eval(getattr(mod.resolveName('lang'), 'value') or '') == 'fr'
+
+@systemcls_param
+def test_constant_class(systemcls: Type[model.System]) -> None:
+    """
+    It can recognize constants define as class members.
+    """
+    mod = fromText('''
+    class Clazz:
+        """Class."""
+        LANG = 'FR'
+    ''', systemcls=systemcls)
+    assert getattr(mod.resolveName('Clazz.LANG'), 'kind') == model.DocumentableKind.CONSTANT
+    assert ast.literal_eval(getattr(mod.resolveName('Clazz.LANG'), 'value') or '') == 'FR'
+
+
+@systemcls_param
+def test_constant_init(systemcls: Type[model.System]) -> None:
+    """
+    It can recognize constants define as instance members.
+    """
+    mod = fromText('''
+    class Clazz:
+        """Class."""
+        def __init__(**args):
+            self.LANG = 'FR'
+    ''', systemcls=systemcls)
+    assert getattr(mod.resolveName('Clazz.LANG'), 'kind') == model.DocumentableKind.CONSTANT
+    assert ast.literal_eval(getattr(mod.resolveName('Clazz.LANG'), 'value') or '') == 'FR'
```

Comment:
Should we actually consider immutable instance variables to be constants?

The convention of constants with all-caps names originates from languages that have compile-time constants, like C and Java. If we'd translate that to Python, a constant would characterized by being both immutable and global.

If we instead define a constant as an immutable named value, then instances could have constants too. We'd need to recognize a whole lot of additional code constructs though to handle this consistently. For example, all attributes in a class decorated with `@attr.s(frozen=True)` would be constants. And perhaps even read-only properties should be considered constants then.

I think it would be simplest to for now only have globals as constants and later decide how we want to document immutable (read-only) instance attributes (both variables and properties).

Reply:
I understand, I kindda agree too.

Reply:
I've fixed this now. There is now 2 different warnings depending on the situation.